pkg/cephfs/volume.go

```diff
@@ -17,140 +17,109 @@ limitations under the License.
 package cephfs
 
 import (
-	"fmt"
-	"os"
-	"path"
+	"strconv"
+	"strings"
+
+	"github.com/ceph/ceph-csi/pkg/util"
 
 	"k8s.io/klog"
 )
 
 const (
-	cephVolumesRoot = "csi-volumes"
+	namespacePrefix = "fsvolumens_"
+)
 
-	namespacePrefix = "ns-"
+var (
+	cephfsInit = false
 )
 
-func getCephRootPathLocal(volID volumeID) string {
-	return fmt.Sprintf("%s/controller/volumes/root-%s", PluginFolder, string(volID))
-}
+func getVolumeRootPathCeph(volOptions *volumeOptions, cr *credentials, volID volumeID) string {
+	stdout, _, err := util.ExecCommand(
+		"ceph",
+		"fs",
+		"subvolume",
+		"getpath",
+		volOptions.FsName,
+		string(volID),
+		"--group_name",
+		"csi",
+		"-m", volOptions.Monitors,
+		"-c", util.CephConfigPath,
+		"-n", cephEntityClientPrefix+cr.id,
+		"--key="+cr.key)
 
-func getCephRootVolumePathLocal(volID volumeID) string {
-	return path.Join(getCephRootPathLocal(volID), cephVolumesRoot, string(volID))
-}
-
-func getVolumeRootPathCeph(volID volumeID) string {
-	return path.Join("/", cephVolumesRoot, string(volID))
+	if err != nil {
+		klog.Errorf("failed to get the rootpath for the vol %s(%s)", string(volID), err)
+		return ""
+	}
+	return strings.TrimSuffix(string(stdout), "\n")
 }
 
 func getVolumeNamespace(volID volumeID) string {
 	return namespacePrefix + string(volID)
 }
 
-func setVolumeAttribute(root, attrName, attrValue string) error {
-	return execCommandErr("setfattr", "-n", attrName, "-v", attrValue, root)
-}
-
-func createVolume(volOptions *volumeOptions, adminCr *credentials, volID volumeID, bytesQuota int64) error {
-	if err := mountCephRoot(volID, volOptions, adminCr); err != nil {
-		return err
-	}
-	defer unmountCephRoot(volID)
-
-	var (
-		volRoot         = getCephRootVolumePathLocal(volID)
-		volRootCreating = volRoot + "-creating"
-	)
-
-	if pathExists(volRoot) {
-		klog.V(4).Infof("cephfs: volume %s already exists, skipping creation", volID)
-		return nil
-	}
-
-	if err := createMountPoint(volRootCreating); err != nil {
-		return err
-	}
-
-	if bytesQuota > 0 {
-		if err := setVolumeAttribute(volRootCreating, "ceph.quota.max_bytes", fmt.Sprintf("%d", bytesQuota)); err != nil {
+func createVolume(volOptions *volumeOptions, cr *credentials, volID volumeID, bytesQuota int64) error {
+	if !cephfsInit {
+		_, _, err := util.ExecCommand(
```

Comment:
if the subvolume group already exists will it pass or return an error?

Reply:
It passes